modules/core/src-jvm/smithy4s/Timestamp.scala

```diff
@@ -0,0 +1,371 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+
+import smithy.api.TimestampFormat
+import smithy4s.Timestamp._
+
+import java.time.Instant
+import scala.util.Try
+import scala.util.control.NoStackTrace
+
+case class Timestamp private(epochSecond: Long, nano: Int) {
+  def isAfter(other: Timestamp): Boolean = {
+    val diff = epochSecond - other.epochSecond
+    diff > 0 || diff == 0 && nano > other.nano
+  }
+
+  def format(format: TimestampFormat): String = format match {
+    case TimestampFormat.DATE_TIME     => formatDateTime
+    case TimestampFormat.EPOCH_SECONDS => formatEpochSeconds
+    case TimestampFormat.HTTP_DATE     => formatHTTPDate
+  }
+
+  /** JVM platform only method */
+  def toInstant: Instant = Instant.ofEpochSecond(epochSecond, nano.toLong)
+
+  override def toString: String = format(TimestampFormat.DATE_TIME)
+
+  /** Has platform specific implementation */
+  private[this] def formatDateTime: String = {
+    val s = new java.lang.StringBuilder(32)
+    val epochDay =
+      (if (epochSecond >= 0) epochSecond
+      else epochSecond - 86399) / 86400 // 86400 == seconds per day
+    val secsOfDay = (epochSecond - epochDay * 86400).toInt
+    var marchZeroDay = epochDay + 719468 // 719468 == 719528 - 60 == days 0000 to 1970 - days 1st Jan to 1st Mar
+    var adjustYear = 0
+    if (marchZeroDay < 0) { // adjust negative years to positive for calculation
+      val adjust400YearCycles = to400YearCycle(marchZeroDay + 1) - 1
+      adjustYear = adjust400YearCycles * 400
+      marchZeroDay -= adjust400YearCycles * 146097L
+    }
+    var year = to400YearCycle(marchZeroDay * 400 + 591)
+    var marchDayOfYear = toMarchDayOfYear(marchZeroDay, year)
+    if (marchDayOfYear < 0) { // fix year estimate
+      year -= 1
+      marchDayOfYear = toMarchDayOfYear(marchZeroDay, year)
+    }
+    val marchMonth = (marchDayOfYear * 17135 + 6854) >> 19 // (marchDayOfYear * 5 + 2) / 153
+    year += (marchMonth * 3277 >> 15) + adjustYear // year += marchMonth / 10 + adjustYear (reset any negative year and convert march-based values back to january-based)
+    val month = marchMonth +
+      (if (marchMonth < 10) 3
+      else -9)
+    val day = marchDayOfYear - ((marchMonth * 1002762 - 16383) >> 15) // marchDayOfYear - (marchMonth * 306 + 5) / 10 + 1
+    append4Digits(year, s)
+    append2Digits(month, s.append('-'))
+    append2Digits(day, s.append('-'))
+    appendTime(secsOfDay, s.append('T'))
+    s.append('Z').toString
+  }
+
+  /** Has platform specific implementation */
+  private[this] def formatEpochSeconds: String = {
+    val s = new java.lang.StringBuilder(32)
+    s.append(epochSecond)
+    appendNano(nano, s)
+    s.toString
+  }
+
+  /** Has platform specific implementation */
+  private[this] def formatHTTPDate: String = {
```

Comment:
out of curiosity, what makes it platform specific ? From a quick look, it seems like it would cross-compile, so I presume it's a case that this implementation would not be optimal on JS. 

If I want to cross-publish to Native, will I be able to share the sources between JVM and Native ?

Reply:
Yes, the current draft is not optimized for any of platforms. 

I'm going to have own Timestamp sources for platforms.

If JVM version should run on JDK 8 then it will be the same as for Scala Native.